### ISSUE TITLE
Spread operator custom function called multiple times

### ISSUE THREAD

**Issue by jackzard** [solidjs/solid#1321]
### Describe the bug

when call spread with custom function, the function is called multiple times, when using version before 1.6.0 it's only run once.

### Your Example Website or App

https://playground.solidjs.com/?version=1.5.9#NobwRAdghgtgpmAXGGUCWEwBowBcCeADgsrgM4Ae2YZA9gK4BOAxiWGjIbY7gAQiM4EACZxGAX14AzRrRi8AOjVoAbNMIC0AKzIB6AO5wARkoUQOXHv3iMA5nAAKswmUky5i5Ws07T5ztx8IMyCULhwAMq43HBusvJKdN7aemTRgn4WgfwOgVAqcR4A5EnqKQbGRWZmUvQQzLhotBC8AMIMEOGMABQAlPxmvEO8zM1pvMBpMQC6vAC8I6HhUTHdwSpQZGQAMmhpiPxQB7iM9LHivYPDoxDjtEZa8wMtw69khKHCB33zAHzPr0B1zGqjgADoVLRbN0ilIWsx8io4MIipcXkDXoJcEwWjZ7E5aC5ulNBFgARiKYIpAc4H8RiCkRCod04GiKa8LlcgeIubweRBeVicbxurzXgAeYRoABu-DB8vuWjB70+fXEul+vLR-LMghEYm6Pzm-3F7TqXQ1ZOEtGY9HgnTB9lwAFEkfbcAAhfAASWEorAUEIhCUvQAhGiwOJpkA

### Steps to Reproduce the Bug or Issue

1. Visit playground
2. Check console

### Expected behavior

expected function only run once

### Platform

- OS: [e.g. macOS, Windows, Linux]
- Browser: [e.g. Chrome, Safari, Firefox]
- Version: [e.g. 91.1]

**Comment by Brendan-csel** [comment]
Can confirm we're seeing this too ...with our `Box` component that is used *everywhere*.

**Comment by lxsmnsyc** [comment]
before 1.6.0 I think similar behavior also happened (since this was already reported before).

**Comment by Brendan-csel** [comment]
Here is a more mimimal reproduction

For 1.5.9 https://playground.solidjs.com/?hash=237697647&version=1.5.9

For 1.6.1 https://playground.solidjs.com/?hash=2093492072&version=1.6.1

**Comment by lxsmnsyc** [comment]
> Here is a more mimimal reproduction
> 
> For 1.5.9 https://playground.solidjs.com/?hash=237697647&version=1.5.9
> 
> For 1.6.1 https://playground.solidjs.com/?hash=2093492072&version=1.6.1

Okay 4 times is definitely a bug

**Comment by ryansolid** [comment]
This isn't a bug but it also isn't great.

We merged spreads now so the logic is a bit different. I was thinking I wouldn't need to Memo.. I hate taking that out of the end user's hands, but what we are seeing here is every object access calling the function again. `get`, `has`, even `keys` to iterate over them in the `spread`. It involves resolving the function every time. Any cache we implemented would get invalidated by any reactive change so we really just want `createMemo` here to make it not per property access. Signal or Memo as a source won't have this problem so in a sense it isn't unlike how we group effects but I admit this can be much more than people would expect.

**Comment by Brendan-csel** [comment]
Tricky.  I know we use spreads **a lot**.

It seems even the simplest usage will "call back" up to the sources at least 4 times. https://playground.solidjs.com/?hash=1667773200&version=1.6.1

**Comment by lxsmnsyc** [comment]
@Brendan-csel tbf 1.5 spread evaluates the source function as many times as any prop of it is accessed if it spread into a component.

**Comment by LiQuidProQuo** [comment]
**A.** This is the only case you can avoid mergeProps with props spreading `{...props}`
a single spread, without any dynamic expression or additional attributes.
```html
<div {...allPropsComesFromHereObj} />
```
```js
spread(_el$, allPropsComesFromHereObj, false, false);
```



**B.** The moment an expression or a prop is added to an element with prop spreading
all props are passed through `mergeProps` logic.

for the object returned by **mergeProps**, any key accessed, search all the sources until the key is found.
if a source is a function, it is called.


```html
<div {...allPropsComesFromHereFn()} />
```
```js
spread(_el$, mergeProps(allPropsComesFromHere), false, false);
```

---- 

 - there is only 1 way to avoid mergeProps logic with a spread. ( see case **A** )
 - if you use a function in the spread, assume it will be called multiple times ( due to internal key lookup when applying props)
 - avoid functions in spread that are expensive to compute without Memoization 
 - avoid createMemo.

**Comment by Brendan-csel** [comment]
> I think the best course here is to wrap the spread source with `memo`

Case-by-case in app code?  Or within spread?  (trade-offs)

**Comment by lxsmnsyc** [comment]
> Case-by-case in app code? Or within spread? (trade-offs)

In reactive spreads, this would prevent further re-evaluation since the first call is already memoized. Reactive preads are pretty counter-behavior because it re-evaluates on every prop access, and on top of that, the issue above.

**Comment by Brendan-csel** [comment]
Doing some testing with combinations of version 1.5.9 and 1.6.1 and with our suspect spread memoized and not.  

The following is scripting time for our "design system" demo/test page which is pretty large (has examples of all our components).

| Version   | Memo | Scripting time |
| --------- | ------------ | ---- |
| 1.5.9 | No | ~195ms |
| 1.5.9 | Yes | ~165ms |
| 1.6.1 | No | ~150ms |
| 1.6.1 | Yes | ~145ms |

So for practical purposes 1.6.1 seems much fastest regardless (nice work @ryansolid) - and memoizing the suspect spread had less impact than with 1.5.9.

The times above are approximately where they settled after about 10 runs in my Performance tab - so not particularly scientific, but repeatable as I swapped between versions several times since difference seemed so significant.

**Comment by lxsmnsyc** [comment]
that's good to see

**Comment by ryansolid** [comment]
Still debating automemo because of the number of hits and the confusion. But appreciate the numbers.

**Comment by LiQuidProQuo** [comment]
@Brendan-csel  those are interesting numbers. 

I am seeing **1.5** faster than **1.6** 
in "raw" performance.

10,000 elements. 
```js
<div {...spread()} />;
``` 

https://playground.solidjs.com/?hash=1407294866&version=1.5.9
![image](https://user-images.githubusercontent.com/105608035/199144757-08541eae-6e6e-40cd-b53d-9ea43c4dc12c.png)


https://playground.solidjs.com/?hash=-1559231175&version=1.6.1
![image](https://user-images.githubusercontent.com/105608035/199144818-953cc4c3-09e9-40e4-a686-bb5340f3f031.png)

**Comment by Brendan-csel** [comment]
Does the playground run in dev mode?  I was doing full builds.

**Comment by LiQuidProQuo** [comment]
yes unfortunately, there is no dev/production toggle support in playground yet.

**Comment by ryansolid** [comment]
It is interesting to me this is noticed now. Because we never memoized this. I guess we nested-wrapped it though. In that we created an effect if it was a function and then had that create an effect that wrapped the actual spreading.
